app/routes.py

```diff
@@ -1,2 +1,286 @@
+        return make_response("", 404)
+
+    if request.method == "GET":
+        if not task.completed_at:
+            if task.goal_id:
+                response_value = {"task":{
+                    "id": task.task_id,
+                    "goal_id": task.goal_id,
+                    "title": task.title,
+                    "description": task.description,
+                    "is_complete": False
+                }}
+            else:
+                response_value = {"task":{
+                    "id": task.task_id,
+                    "title": task.title,
+                    "description": task.description,
+                    "is_complete": False
+                }}
+
+            return make_response(response_value, 200)
+        else:
+            response_value = {"task":{
+                "id": task.task_id,
+                "title": task.title,
+                "description": task.description,
+                "is_complete": task.completed_at
+            }}
+
+            return make_response(response_value, 200)
+    elif request.method == "PUT":
+        form_data = request.get_json()
+
+        task.title = form_data["title"]
+        task.description = form_data["description"]
+        if form_data.get("completed_at"):
+            task.completed_at = datetime.now()
+
+        db.session.commit()
+
+        if not task.completed_at:
+            response_value = {"task":{
+                "id": task.task_id,
+                "title": task.title,
+                "description": task.description,
+                "is_complete": False
+            }}
+
+            return make_response(response_value, 200)
+        else:
+            response_value = {"task":{
+                "id": task.task_id,
+                "title": task.title,
+                "description": task.description,
+                "is_complete": True
+            }}
+            return make_response(response_value, 200)
+
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        return make_response({"details": f'Task {task.task_id} "{task.title}" successfully deleted'},200)
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return make_response("", 404)
+
+    task.completed_at = datetime.now()
+
+    response_value = {"task":{
+                "id": task.task_id,
+                "title": task.title,
+                "description": task.description,
+                "is_complete": True
+    }}
+    url= 'https://slack.com/api/chat.postMessage'
+    header_values = {'AUTHORIZATION': os.environ.get("AUTHORIZATION")}
+    slack_values = {"text" : f"Someone just completed the task {task.title}",
+                    "channel" : "task list api"
+    }
+    requests.post(url, headers=header_values, params=slack_values)
+    db.session.commit()
+
+    return (make_response(response_value, 200))
```

Comment:
There are two parts to this function: (1) marking the task as complete in our database and (2) pinging the slackbot. To practice the single responsibility principle, we can move the logic for the slack bot into a helper function. This will also allow the slack bot to be used by other routes too (ex: confirming post, put, or get requests, etc.)